analysis/R/assoc.R

```diff
@@ -0,0 +1,164 @@
+  # cohort
+  # - a combined map (rmap) that collapses cohorts
+  m1 <- lapply(1:params$m, function(z) {
+    m <- sparseMatrix(c(1), c(2), dims = c(1, 2))
+    colnames(m) <- c("FALSE", "TRUE")
+    m
+  })
+  m2 <- sparseMatrix(1:params$m, rep(2, params$m))
+  colnames(m2) <- colnames(m1[[1]])
+  map[[2]]$map <- m1
+  map[[2]]$rmap <- m2
+  
+  # Reports must be of the format
+  #     client name, cohort no, rappor bitstring 1, rappor bitstring 2, ...
+  # Reading to temporary data frame
+  tmp_reports_df <- read.csv(inp$reports,
+                             colClasses = c("character", "integer",
+                                            rep("character", inp$numvars)),
+                             header = TRUE)
+  # Ignore the first column
+  tmp_reports_df <- tmp_reports_df[,-1]
+  
+  # TODO(pseudorandom): Do not assume var 2 is always boolean
+  # Fix which params$k is set to 1 based on which var is boolean
+  # params now is a tuple of params in the standard format as required by
+  # ComputeDistributionEM
+  params = list(params, params)
+  params[[2]]$k = 1
+  
+  # Parsing tmp_reports_df
+  # ComputeDistributionEM allows for different sets of cohorts
+  # for each variable. Here, both sets of cohorts are identical
+  tmp_cohorts <- as.list(tmp_reports_df[1])$cohort
+  tmp_cohorts <- tmp_cohorts + 1  # 1 indexing for EM algorithm
+  # Expanded list of report cohorts
+  cohorts <- rep(list(tmp_cohorts), inp$numvars)
+  # Parse reports from tmp_reports_df cols 2, 3, ...
+  # into a reports array
+  reports <- lapply(1:inp$numvars, function(x) as.list(tmp_reports_df[x + 1]))
+  
+  # Split ASCII strings into array of numerics (as required by assoc analysis)
+  reports <- lapply(1:inp$numvars, function(i) {
+    # apply the following function to each of reports[[1]] and reports[[2]]
+    # TODO(pseudorandom): Do not assume var 2 is always boolean in [[1]]
+    lapply(reports[[i]][[1]], function(x) {
+      # function splits strings and converts them to numeric values
+      # rev needed for endianness
+      rev(as.integer(strsplit(x, split = "")[[1]]))
+    })
+  })
+  
+  # ignore_other = FALSE because in doing association with Basic RAPPOR, we're
+  # going to use the Other category to estimate no. of reports where the bit
+  # is unset
+  # marginals and estimate_var disabled
+  # verbose outputs timing information for efficiency analysis; set to inp$time
+  joint_dist <- ComputeDistributionEM(reports, cohorts, map,
+                                      ignore_other = FALSE,
+                                      params, marginals = NULL,
+                                      estimate_var = FALSE,
+                                      verbose = inp$time)
+  em <- joint_dist$fit
+  time_taken <- proc.time() - ptm
+  cat("Came here \n\n")
+  # Replace Other column name with FALSE. In Basic RAPPOR we use Other
+  # category to estimate no. of reports with unset bits (which are essentially
+  # FALSE)
```

Comment:
This seems hacky, or at least confusing to read.  Can we change the association code to support booleans properly?  False and "other" are two different concepts.